sdk/src/test/java/io/horizen/utils/ChaChaPrngSecureRandomTest.java

```diff
@@ -5,60 +5,269 @@
 import static org.junit.Assert.*;
 
 public class ChaChaPrngSecureRandomTest {
+    static int littleEndianToInt(byte[] data) {
+        return (data[0] & 0xff) | ((data[1] & 0xff) << 8) | ((data[2] & 0xff) << 16) | ((data[3] & 0xff) << 24);
+    }
+
+    static double log2(double v) {
+        return Math.log(v) / Math.log(2);
+    }
+
+    static boolean shannon(byte[] data) {
+        // See Donald Knuth "A mathematical theory of communication"
+        // p. 379-423
+        final double THRESHOLD = 7.9;
+        int[] buckets = new int[256];
+        for (byte b : data) {
+            int i = b & 0xff;
+            buckets[i]++;
+        }
+        double length = data.length;
+        double entropy = 0;
+        for (int i : buckets) {
+            double probability = ((double)i) / length;
+            if (probability > 0.0) {
+                entropy += probability * log2((1.0/probability));
+            }
+        }
+        return entropy >= THRESHOLD;
+    }
+
+    static boolean mean(byte[] data) {
+        // See Donald Knuth "A mathematical theory of communication"
+        // p. 379-423
+        final double THRESHOLD = 126;
+        int[] buckets = new int[256];
+        for (byte b : data) {
+            int i = b & 0xff;
+            buckets[i]++;
+        }
+        double length = data.length;
+        double sum = 0;
+        for (int i = 0; i < buckets.length; i++) {
+            sum += ((double)i) * ((double)buckets[i]);
+        }
+        sum /= length;
+        return sum >= THRESHOLD;
+    }
+
+    // See <https://www.geeksforgeeks.org/estimating-value-pi-using-monte-carlo>
+    static boolean monteCarlo(byte[] data) {
```

Comment:
Could you better explain what is the purpose of this test? Obtaining a good estimation of pi would imply that the sequence of generated random bytes is random enough, because otherwise Monte carlo simulation would fail? Doesn't this heavily depend on the number of rounds being performed?

Reply:
Ok, could you add a comment to briefly explain it? I am not sure how to measure if the approximation range you put in the test is good enough given the number of iterations being performed, but I guess it is ok for a simple test